### PR TITLE
Build properties plugin

### PR DESCRIPTION
**Pull request by mr-archano**
This PR introduces a custom Gradle plugin that can be used to easily integrate `Properties` support in an Android build script. The features are listed in the [`README.md`](https://github.com/novoda/spikes/blob/build_properties_plugin/BuildPropertiesPlugin/README.md) and a sample app is provided to showcase [real usage](https://github.com/novoda/spikes/blob/build_properties_plugin/BuildPropertiesPlugin/sample/app/build.gradle).

### CODE REVIEW

**Comment by stefanhoth** [comment]
This is great! Only read the Readme until now but it's super useful. 

Not intending to bikeshed but the project name should include Android to avoid confusion when we release it publicly.

File: BuildPropertiesPlugin/README.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+To build the latest version of the plugin, run from the root of the repo:
+```
+./gradlew build
+```
+
+The plugin at the moment is not deployed in any public maven repo, therefore your only option is to deploy it to a local one.
+A default setup with an embedded repo (pointing to `.gradle/repo`) is provided, just run the following command from the root
+ of the project:
+```
+./gradlew -Plocal uploadArchives
```
Comment (blundell) [inline BuildPropertiesPlugin/README.md]: you can still use the bintray release plugin to deploy locally, altho it is a bit more wordy than this

**Comment by blundell** [comment]
👍

**Comment by mr-archano** [comment]
@stefanhoth this repo is not in its final shape yet. I would like to keep the general name and add support for plain java projects too (following the work started by @frapontillo a while ago).

File: BuildPropertiesPlugin/README.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    defaultConfig {
+        ...
+        resValueProperty 'api_key', from(buildProperties.secrets), 'apiKey'
+        ...
+    }
+```
+You can omit the name of the resource and let the plugin generate one for you from the property key, enforcing snake casing over camel casing if necessary.
+
+
+#### 5. Load signing coinfiguration from properties
```
Comment (frapontillo) [inline BuildPropertiesPlugin/README.md]: "Configuration"

File: BuildPropertiesPlugin/README.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+|`buildConfigInt` | `buildConfigInt 'TEST_INT', 42`|
+|`buildConfigLong` | `buildConfigLong 'TEST_LONG', System.currentTimeMillis()`|
+|`buildConfigDouble` | `buildConfigDouble 'TEST_DOUBLE', Math.PI`|
+|`buildConfigString` | `buildConfigString 'TEST_STRING', "whateva"`|
+|`resValueInt`| `resValueInt 'debug_test_int', 100`|
+|`resValueBoolean` | `resValueBoolean 'debug_test_bool', true`|
+|`resValueString` | `resValueString 'debug_test_string', 'dunno bro...'`|
+
+#### More on load properties
+When you need to specify a properties file to retrieve a property from you can use the convenience method `from()` referencing one of the files listed in `buildProperties`. If instead you want to use a different file that is not listed there you can use `load()`, specifying the path of the file you want to use. This in turn will lazy-load the properties file and caching it if any further calls in the build script.
```
Comment (frapontillo) [inline BuildPropertiesPlugin/README.md]: Not super clear on how `load` differs from `from` after reading this. Maybe an example?

File: BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildPropertiesPlugin.groovy [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        if (!isAndroidApp) {
+            throw new GradleException("This plugin can be applied only to an Android application project")
+        }
+
+        project.android.extensions.add('buildProperties', project.container(BuildProperties))
+        def container = project.android.extensions.getByName('buildProperties')
+        project.afterEvaluate {
+            container.all { BuildProperties buildProperties ->
+                Properties properties = new Properties()
+                properties.load(new FileInputStream(buildProperties.file))
```
Comment (frapontillo) [inline BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildPropertiesPlugin.groovy]: I would handle the loading logic inside the `BuildProperties` class, but can be done later.

File: BuildPropertiesPlugin/sample/app/build.gradle [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    targetSdkVersion 23
+    versionCode 1
+    versionName "1.0"
+    buildConfigBoolean 'TEST_BOOLEAN', false
+    buildConfigInt 'TEST_INT', 42
+    buildConfigLong 'TEST_LONG', System.currentTimeMillis()
+    buildConfigDouble 'TEST_DOUBLE', Math.PI
+    buildConfigString 'TEST_STRING', "whateva"
+    buildConfigProperty 'TEST_BUILD_PROPERTY', from(buildProperties.secrets), 'superSecret'
+    buildConfigProperty 'TEST_LOAD_PROPERTY', load(rootProject.file('properties/secrets.properties')), 'superSecret'
```
Comment (frapontillo) [inline BuildPropertiesPlugin/sample/app/build.gradle]: So this is equivalent to
```groovy
buildConfigProperty 'TEST_LOAD_PROPERTY', from(buildProperties.secrets), 'superSecret'
```
?

**Comment by frapontillo** [comment]
This is awesome 💯 
Is there a way to add some tests (even in another PR)? So we can refactor and extend easily.

File: BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    final String name
+    File file
+    Properties entries
+
+    BuildProperties(String name) {
+        this.name = name
+    }
+
+    String getName() {
```
Comment (tasomaniac) [inline BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy]: Just a styling question. These getters and setters are not necessary in Groovy. Did you want them to be explicit?

File: BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildPropertiesPlugin.groovy [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                .toLowerCase()
+    }
+
+    private void addSigningConfigSupportTo(target, Project project) {
+        target.ext.signingConfigProperties = { Closure<BuildProperties> getBuildProperties ->
+            project.afterEvaluate {
+                BuildProperties buildProperties = getBuildProperties()
+                Properties properties = buildProperties.entries
+                target.storeFile new File(buildProperties.parentFile, properties.storeFile)
+                target.storePassword properties.storePassword
```
Comment (tasomaniac) [inline BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildPropertiesPlugin.groovy]: The property file needs to have these values, right. Is it obvious or should we have check, or should we have explanation in the readme?

File: BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy (hunk 6 above)
Comment (mr-archano) [inline BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy]: I want to enforce only getters. `file()` as method is needed to achieve fluent API in the `buildProperties` configuration.

Comment (tasomaniac) [inline BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy]: But you have setFile and setEntries as well.  Aren't those getter automatically available to the consumer even when they are not defined explicitly.

Comment (mr-archano) [inline BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy]: `BuildProperties` has changed quite a bit. Setters are gone 💃

**Comment by frapontillo** [comment]
👍 by me, let's not over achieve in this PR, this is still a spike and waiting for more improvements.

File: BuildPropertiesPlugin/README.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+The plugin will automatically retrieve all the needed fields from the properties file.
+Note: the path of the keystore file is considered relative to the path of the specified properties file.
+
+## Bonus
+
+#### Typed `buildConfigField`/`resValue`
+The plugin enhances the `buildConfigField` and `resValue` facilities to enforce types.
+To generate a string field in your `BuildConfig` you used to write:
+```
+buildConfigField 'String', 'LOL', '\"sometimes the picture takes\"'
```
Comment (xrigau) [inline BuildPropertiesPlugin/README.md]: typo: `takes` -> `take` as per https://github.com/novoda/image-loader/issues/166

**Comment by mr-archano** [comment]
@frapontillo after talking with you I looked at what you were doing and laid out the groundwork to have the plain java support you're looking for. I hope to implement the constants class generation soon in a follow-up PR.

File: BuildPropertiesPlugin/README.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+```
+but now you can instead write:
+```
+buildConfigString 'LOL', 'sometimes the picture takes'
```
Comment (xrigau) [inline BuildPropertiesPlugin/README.md]: same as https://github.com/novoda/spikes/pull/109/files#r68730769

File: BuildPropertiesPlugin/README.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+the `-Plocal` flag at build time as explained above (Fear not, Bintray deploy is coming soon).
+
+#### 2. Apply plugin
+The plugin needs to be applied to your Android module after the Android gradle plugin. In your `build.gradle` you should put
+```
+apply plugin: 'build-properties'
+```
+#### 3. List properties files
+In your android project you can add a `buildProperties` configuration listing all the properties files you intend
+to reference in your `android` configuration, eg:
```
Comment (tasomaniac) [inline BuildPropertiesPlugin/README.md]: Documentation still says that this should be defined in `android` configuration.

File: BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy (hunk 6 above)
Comment (tasomaniac) [inline BuildPropertiesPlugin/plugin/src/main/groovy/com/novoda/buildproperties/BuildProperties.groovy]: Not get me wrong 😄 I just wanted to understand your personal opinion on this 👍

File: BuildPropertiesPlugin/README.md (hunk 10 above)
Comment (mr-archano) [inline BuildPropertiesPlugin/README.md]: It actually states you might reference the properties in `buildProperties` configuration in the android configuration (ie: `buildConfigProperty buildProperties.secrets['foo']`)

Comment (tasomaniac) [inline BuildPropertiesPlugin/README.md]: Oh I see. Sorry. If someone reads it for the first time, I think it is clear.

**Comment by xrigau** [comment]
3rd 👍 time to :shipit:

**Pull request merged**